File: sample/src/main/java/com/fondesa/lyra/sample/MainActivity.java
package com.fondesa.lyra.sample;

import android.os.Build;
import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.fondesa.lyra.Lyra;
import com.fondesa.lyra.annotation.SaveState;
import com.fondesa.lyra.coder.gson.base.DefaultGsonCoder;
import com.fondesa.lyra.sample.model.AutoSaveEditText;
import com.fondesa.lyra.sample.model.Model;
import com.fondesa.lyra.sample.model.ParcelableModel;

import java.util.ArrayList;

public class MainActivity extends BaseMainActivity {
    private TextView mTextView;

    @SaveState(DefaultGsonCoder.class)
    private String mText;

    @SaveState
    ParcelableModel mModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        int defaultPadding = getResources().getDimensionPixelSize(R.dimen.default_inner_padding);
        root.setPadding(defaultPadding, defaultPadding, defaultPadding, defaultPadding);

        mTextView = new TextView(this);
        root.addView(mTextView, new ViewGroup.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT));

        Button button = new Button(this);
        button.setText(R.string.change_value);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mText = (mText == null) ? "first" : null;
                mByte = (mByte == null) ? (byte) 1 : null;
                mPrimitiveByte = (mPrimitiveByte == 0) ? (byte) 1 : 0;
                ArrayList<Model> modelList = new ArrayList<>();
                modelList.add(new Model(1, "first"));
                modelList.add(new Model(2, "second"));
                modelList.add(new Model(3, "third"));

                mModelList = (mModelList == null) ? modelList : null;
                mModel = (mModel == null) ? new ParcelableModel(5, "fifth") : null;
                printInfo();
            }
        });
        LinearLayout.LayoutParams buttonParams =
                new LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT);
        buttonParams.gravity = Gravity.BOTTOM | Gravity.CENTER_HORIZONTAL;

        root.addView(button, buttonParams);

        AutoSaveEditText autoSaveEditText = new AutoSaveEditText(this);
        autoSaveEditText.setId(R.id.auto_save_edit_text);
        root.addView(autoSaveEditText, new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));

        setContentView(root, new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.ICE_CREAM_SANDWICH) {
            Lyra.instance().restoreState(this, savedInstanceState);
        }

        printInfo();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.ICE_CREAM_SANDWICH) {
            Lyra.instance().saveState(this, outState);
        }
    }

    private void printInfo() {
        mTextView.setText(null);

        mTextView.append("String: " + mText);
        appendNl();
        mTextView.append("Primitive byte: " + mPrimitiveByte);
        appendNl();
        mTextView.append("Byte object: " + mByte);
        appendNl();

        mTextView.append("Model list: ");
        if (mModelList != null) {
            String text = "";
            for (int i = 0; i < mModelList.size(); i++) {
                Model m = mModelList.get(i);
                if (i != 0)
                    text += " - ";

                text += m.id + ": " + m.value;
            }
            mTextView.append(text);
        } else {
            mTextView.append("null");
        }

        appendNl();

        mTextView.append("Parcelable model: ");
        if (mModel != null) {
            mTextView.append(mModel.id + ": " + mModel.value);
        } else {
            mTextView.append("null");
        }
    }

    private void appendNl() {
        mTextView.append("\n");
    }
}

File: sample/src/main/kotlin/com/fondesa/lyra/sample/MainActivity.kt
package com.fondesa.lyra.sample

import android.os.Build
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import com.fondesa.lyra.Lyra
import com.fondesa.lyra.annotation.SaveState
import com.fondesa.lyra.coder.gson.base.DefaultGsonCoder
import com.fondesa.lyra.sample.model.AutoSaveEditText
import com.fondesa.lyra.sample.model.Model
import com.fondesa.lyra.sample.model.ParcelableModel
import java.util.*

class MainActivity : BaseMainActivity() {
    private lateinit var textView: TextView

    @SaveState(DefaultGsonCoder::class)
    private var text: String? = null

    @SaveState
    internal var model: ParcelableModel? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val rootView = initView()
        val rootViewParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        setContentView(rootView, rootViewParams)

        restoreState(savedInstanceState)

        printInfo()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.ICE_CREAM_SANDWICH) {
            Lyra.instance().saveState(this, outState)
        }
    }

    private fun initView(): View {
        val root = LinearLayout(this)
        root.orientation = LinearLayout.VERTICAL
        val defaultPadding = resources.getDimensionPixelSize(R.dimen.default_inner_padding)
        root.setPadding(defaultPadding, defaultPadding, defaultPadding, defaultPadding)

        textView = TextView(this)
        root.addView(textView, ViewGroup.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT))

        val button = Button(this)
        button.setText(R.string.change_value)
        button.setOnClickListener {
            text = if (text == null) "first" else null
            randomByte = if (randomByte.toInt() == 0) 1 else 0

            if (modelList == null) {
                modelList = ArrayList<Model>()
                modelList!!.add(Model(1, "first"))
                modelList!!.add(Model(2, "second"))
                modelList!!.add(Model(3, "third"))
            } else {
                modelList = null
            }
            model = if (model == null) ParcelableModel(5, "fifth") else null
            printInfo()
        }
        val buttonParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        buttonParams.gravity = Gravity.BOTTOM or Gravity.CENTER_HORIZONTAL

        root.addView(button, buttonParams)

        val autoSaveEditText = AutoSaveEditText(this)
        autoSaveEditText.id = R.id.auto_save_edit_text
        root.addView(autoSaveEditText, ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT))
        return root
    }

    private fun restoreState(savedInstanceState: Bundle?) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.ICE_CREAM_SANDWICH) {
            Lyra.instance().restoreState(this, savedInstanceState)
        }
    }

    private fun printInfo() {
        textView.text = null

        textView.append("String: $text\n")
        textView.append("Byte: $randomByte\n")

        textView.append("Model list: ")
        if (modelList != null) {
            var text = ""
            for (i in 0..modelList!!.size - 1) {
                if (i != 0)
                    text += " - "

                val m = modelList!![i]
                text += "${m.id}: ${m.value}"
            }
            textView.append(text)
        } else {
            textView.append("null")
        }

        textView.append("\n")

        textView.append("Parcelable model: ")
        if (model != null) {
            textView.append("${model!!.id}: ${model!!.value}")
        } else {
            textView.append("null")
        }
    }
}